dev/scripts/build.ts

```diff
@@ -3,46 +3,139 @@ import { join } from "node:path"
+    attest: {
+        pattern: /"(\.\.\/)+[^"]*attest\/[^"]*\.js"/g,
+        replacement: `"@arktype/attest"`
+    },
+    utils: {
+        pattern: /"(\.\.\/)+[^"]*utils\/[^"]*\.js"/g,
+        replacement: `"@arktype/utils"`
+    }
+} as const
+
+const ignorePaths = [
+    "node_modules",
+    "package.json",
+    `${tempTsConfigBaseName}.tsbuildinfo`
+]
+
+const fixBuildPaths: (buildPath: string) => void = rewritePaths(
+    replacementDictionary,
+    ignorePaths
+)
+
+const buildFormat = (module: ModuleKind) => {
+    const moduleKindDir = ModuleKindToDir[module]
+    const outDir = join(outRoot, moduleKindDir)
+    const utilsSrc = join(outDir, ...Sources.utils, "src")
+    const attestSrc = join(outDir, ...Sources.attest, "src")
+    const utilsTarget = join(
+        packageRoot,
+        ...Sources.utils,
+        "dist",
+        moduleKindDir
+    )
+    const attestTarget = join(
+        packageRoot,
+        ...Sources.attest,
+        "dist",
+        moduleKindDir
+    )
+
+    const tempTsConfig = {
+        ...baseTsConfig,
+        include: ["src", Sources.utils.join("/"), Sources.attest.join("/")],
         compilerOptions: {
-            ...tsConfigData.compilerOptions,
+            ...compilerOptions,
             noEmit: false,
             module,
             outDir
         }
+    }
+
+    const writePackageManifest = writeManifest({
+        type: ModuleKindToPackageType[module]
     })
+
+    writeJson(tempTsConfigPath, tempTsConfig)
+
     try {
-        shell(`pnpm tsc --project ${tempTsConfig}`)
+        shell(`pnpm tsc --project ${tempTsConfigPath}`)
         const outSrc = join(outDir, "src")
+        const outDev = join(outDir, "dev")
         // not sure which setting to change to get it to compile here in the first place
-        cpSync(outSrc, outDir, {
-            recursive: true,
-            force: true
-        })
-        if (module === "commonjs") {
-            writeJson(join(outDir, "package.json"), {
-                type: "commonjs"
-            })
-        }
-        rmSync(outSrc, { recursive: true, force: true })
+        clone(outSrc, outDir)
+        clone(utilsSrc, utilsTarget)
+        clone(attestSrc, attestTarget)
+
+        writePackageManifest(repoDirs.root, outDir)
+        writePackageManifest(repoDirs.attest, attestTarget)
+        writePackageManifest(repoDirs.utils, utilsTarget)
+
+        nuke(outSrc)
+        nuke(outDev)
+        nuke(utilsSrc)
+        nuke(attestSrc)
+
+        fixBuildPaths(outDir)
+        fixBuildPaths(attestTarget)
+        /**
+         * We don't need to rewrite any of the paths in `dev/utils/dist` at the
+         * moment, since it doesn't (currently) depend on any local packages
+         */
+        // fixBuildPaths(utilsTarget)
     } finally {
-        rmSync(tempTsConfig, { force: true })
+        rmSync(tempTsConfigPath, { force: true })
     }
 }
 
+type ModuleKind = (typeof ModuleKind)[keyof typeof ModuleKind]
+const ModuleKind = {
```

Comment:
I appreciate the diligence, but in general I'd probably just use a string union here. You can then define the mapped types as `... satisfies Record<ModuleKind, string>`. We also don't generally capitalize the names of config objects

I'm sure others would feel differently, so definitely more a stylistic consistency thing!

Reply:
I can remove these, but to give you my reasoning:

You can think of these 3 dictionaries as 3 total functions; by re-using the identifiers from one to the next, I'm ensuring that they can continue to be composed with one another.

It was helpful for me to do this because I was getting confused. For example, when using a string union "commonjs" | "esnext", it's not immediately clear that these strings belong to the keys in this enum:

```typescript
enum ModuleKind {
    None = 0,
    CommonJS = 1,
    AMD = 2,
    UMD = 3,
    System = 4,
    ES2015 = 5,
    ES2020 = 6,
    ES2022 = 7,
    ESNext = 99,
    Node16 = 100,
    NodeNext = 199
 }
```

In other words, that makes the functions partial (as opposed to total).

Then, we were re-using one of the values ("commonjs") for the "type" field of the `cjs/package.json` (and using "module" for the other).

All of this was getting a little difficult to keep track of, so I went ahead created these to make the mapping between these configs explicit.

Would it make more sense if I used `commonjs` and `module` as the keys, and derived the other enums from those?

Reply:
I can remove these, but to give you my reasoning:

You can think of these 3 dictionaries as 3 total functions; by re-using the identifiers from one to the next, I'm ensuring that they can continue to be composed with one another.

It was helpful for me to do this because I was getting confused. For example, when using a string union "commonjs" | "esnext", it's not immediately clear that these strings belong to the keys in this enum:

```typescript
enum ModuleKind {
    None = 0,
    CommonJS = 1,
    AMD = 2,
    UMD = 3,
    System = 4,
    ES2015 = 5,
    ES2020 = 6,
    ES2022 = 7,
    ESNext = 99,
    Node16 = 100,
    NodeNext = 199
 }
```

In other words, that makes the functions partial (as opposed to total).

Then, we were re-using one of the values ("commonjs") for the "type" field of the `cjs/package.json` (and using "module" for the other).

All of this was getting a little difficult to keep track of, so I went ahead created these to make the mapping between these configs explicit.

Would it make more sense if I used `commonjs` and `module` as the keys, and derived the other enums from those?

Reply:
Personally, I just don't use enums or enum-like objects in the codebase. If I want a set of keys I can check at runtime, I tend to use a Record mapping each key to `true` or a different mapping if there's something useful I can associate them with. I don't usually like to map strings to themselves as it feels duplicative.

It doesn't really matter that TS implements `ModuleKind` as an enum under the hood, because what we're interested in is the string that populates the config file, not the underlying representation once it's been parsed.

I can see how as written, it might not be very clear what the literals  "commonjs" | "esnext" correlate with. They way I'd normally address this is by giving them an alias like `TsConfigModuleFormat` and using that where I expect one of those two values. Another approach would be to have two keys like "cjs" and "esm" or similar, each of which map to a context object for that build format. Something like:

```ts
type BuildFormatContext = {
    tsConfigModule: TsConfigModuleKind
    packageJsonType: PackageJsonModuleKind
}
```

This could also be expanded if we wanted to add other context like an explicit extension in the future.

Just to reiterate, I don't think there's anything wrong with the approach you chose, it's just not the way the situation is typically handled in the repository so I'd prefer using one of the other approaches.